In [27]:
!pip install geopy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 kB 1.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.4/125.4 kB 4.6 MB/s eta 0:00:00


In [29]:
import pandas as pd
import os
from geopy.distance import geodesic

In [2]:
file_names = ["./Data/predict-energy-behavior-of-prosumers/client.csv", "./Data/predict-energy-behavior-of-prosumers/county_id_to_name_map.json", "./Data/predict-energy-behavior-of-prosumers/electricity_prices.csv", "./Data/predict-energy-behavior-of-prosumers/forecast_weather.csv", "./Data/predict-energy-behavior-of-prosumers/gas_prices.csv", "./Data/predict-energy-behavior-of-prosumers/historical_weather.csv", "./Data/predict-energy-behavior-of-prosumers/train.csv", "./Data/predict-energy-behavior-of-prosumers/weather_station_to_county_mapping.csv"]

<h2>Read files to dataframes </h2>

In [3]:
 os.getcwd()

'/Users/melahattayli/projects'

In [4]:
os.listdir('./Data/predict-energy-behavior-of-prosumers')

['enefit',
 'client.csv',
 'weather_station_to_county_mapping.csv',
 'gas_prices.csv',
 'forecast_weather.csv',
 'electricity_prices.csv',
 'example_test_files',
 'county_id_to_name_map.json',
 'public_timeseries_testing_util.py',
 'train.csv',
 '.ipynb_checkpoints',
 'historical_weather.csv']

In [5]:
train_df = pd.read_csv("./Data/predict-energy-behavior-of-prosumers/train.csv")
train_df

county  is_business  product_type   target  is_consumption  \
0             0            0             1    0.713               0   
1             0            0             1   96.590               1   
2             0            0             2    0.000               0   
3             0            0             2   17.314               1   
4             0            0             3    2.904               0   
...         ...          ...           ...      ...             ...   
2018347      15            1             0  197.233               1   
2018348      15            1             1    0.000               0   
2018349      15            1             1   28.404               1   
2018350      15            1             3    0.000               0   
2018351      15            1             3  196.240               1   

                    datetime  data_block_id   row_id  prediction_unit_id  
0        2021-09-01 00:00:00              0        0                   0  
1        2021-09-01 00:00:00              0        1                   0  
2        2021-09-01 00:00:00              0        2                   1  
3        2021-09-01 00:00:00              0        3                   1  
4        2021-09-01 00:00:00              0        4                   2  
...                      ...            ...      ...                 ...  
2018347  2023-05-31 23:00:00            637  2018347                  64  
2018348  2023-05-31 23:00:00            637  2018348                  59  
2018349  2023-05-31 23:00:00            637  2018349                  59  
2018350  2023-05-31 23:00:00            637  2018350                  60  
2018351  2023-05-31 23:00:00            637  2018351                  60  

[2018352 rows x 9 columns]

In [6]:
client_df =  pd.read_csv("./Data/predict-energy-behavior-of-prosumers/client.csv")
client_df

product_type  county  eic_count  installed_capacity  is_business  \
0                 1       0        108              952.89            0   
1                 2       0         17              166.40            0   
2                 3       0        688             7207.88            0   
3                 0       0          5              400.00            1   
4                 1       0         43             1411.00            1   
...             ...     ...        ...                 ...          ...   
41914             1      15         51              415.60            0   
41915             3      15        161             2035.75            0   
41916             0      15         15              620.00            1   
41917             1      15         20              624.50            1   
41918             3      15         55             2188.20            1   

             date  data_block_id  
0      2021-09-01              2  
1      2021-09-01              2  
2      2021-09-01              2  
3      2021-09-01              2  
4      2021-09-01              2  
...           ...            ...  
41914  2023-05-29            637  
41915  2023-05-29            637  
41916  2023-05-29            637  
41917  2023-05-29            637  
41918  2023-05-29            637  

[41919 rows x 7 columns]

In [7]:
gas_prices_df =  pd.read_csv("./Data/predict-energy-behavior-of-prosumers/gas_prices.csv")
gas_prices_df

forecast_date  lowest_price_per_mwh  highest_price_per_mwh origin_date  \
0      2021-09-01                 45.23                  46.32  2021-08-31   
1      2021-09-02                 45.62                  46.29  2021-09-01   
2      2021-09-03                 45.85                  46.40  2021-09-02   
3      2021-09-04                 46.30                  46.80  2021-09-03   
4      2021-09-05                 46.30                  46.58  2021-09-04   
..            ...                   ...                    ...         ...   
632    2023-05-26                 29.10                  34.10  2023-05-25   
633    2023-05-27                 28.30                  34.10  2023-05-26   
634    2023-05-28                 28.10                  34.10  2023-05-27   
635    2023-05-29                 28.16                  36.98  2023-05-28   
636    2023-05-30                 29.00                  34.00  2023-05-29   

     data_block_id  
0                1  
1                2  
2                3  
3                4  
4                5  
..             ...  
632            633  
633            634  
634            635  
635            636  
636            637  

[637 rows x 5 columns]

In [8]:
forecast_weather_df = pd.read_csv("./Data/predict-energy-behavior-of-prosumers/forecast_weather.csv")
forecast_weather_df 

latitude  longitude      origin_datetime  hours_ahead  temperature  \
0            57.6       21.7  2021-09-01 02:00:00            1    15.655786   
1            57.6       22.2  2021-09-01 02:00:00            1    13.003931   
2            57.6       22.7  2021-09-01 02:00:00            1    14.206567   
3            57.6       23.2  2021-09-01 02:00:00            1    14.844507   
4            57.6       23.7  2021-09-01 02:00:00            1    15.293848   
...           ...        ...                  ...          ...          ...   
3424507      59.7       26.2  2023-05-30 02:00:00           48    10.150049   
3424508      59.7       26.7  2023-05-30 02:00:00           48    10.259790   
3424509      59.7       27.2  2023-05-30 02:00:00           48    10.467798   
3424510      59.7       27.7  2023-05-30 02:00:00           48    10.694116   
3424511      59.7       28.2  2023-05-30 02:00:00           48    11.233179   

          dewpoint  cloudcover_high  cloudcover_low  cloudcover_mid  \
0        11.553613         0.904816        0.019714        0.000000   
1        10.689844         0.886322        0.004456        0.000000   
2        11.671777         0.729034        0.005615        0.000000   
3        12.264917         0.336304        0.074341        0.000626   
4        12.458887         0.102875        0.088074        0.000015   
...            ...              ...             ...             ...   
3424507   5.662744         0.061066        0.257339        0.004059   
3424508   5.922754         0.107880        0.895203        0.198700   
3424509   6.072534         0.000000        0.387466        0.448853   
3424510   7.009058         0.000000        0.505936        0.997314   
3424511   7.069238         0.703461        0.292313        0.934021   

         cloudcover_total  10_metre_u_wind_component  \
0                0.905899                  -0.411328   
1                0.886658                   0.206347   
2                0.730499                   1.451587   
3                0.385468                   1.090869   
4                0.176590                   1.268481   
...                   ...                        ...   
3424507          0.308594                   6.890626   
3424508          0.906036                   6.890626   
3424509          0.530518                   6.214112   
3424510          0.997498                   5.810060   
3424511          0.989716                   2.719728   

         10_metre_v_wind_component  data_block_id    forecast_datetime  \
0                        -9.106137              1  2021-09-01 03:00:00   
1                        -5.355405              1  2021-09-01 03:00:00   
2                        -7.417905              1  2021-09-01 03:00:00   
3                        -9.163999              1  2021-09-01 03:00:00   
4                        -8.975766              1  2021-09-01 03:00:00   
...                            ...            ...                  ...   
3424507                  -3.258144            637  2023-06-01 02:00:00   
3424508                  -3.768032            637  2023-06-01 02:00:00   
3424509                  -4.700405            637  2023-06-01 02:00:00   
3424510                  -4.338345            637  2023-06-01 02:00:00   
3424511                  -0.922939            637  2023-06-01 02:00:00   

         direct_solar_radiation  surface_solar_radiation_downwards  snowfall  \
0                           0.0                                0.0       0.0   
1                           0.0                                0.0       0.0   
2                           0.0                                0.0       0.0   
3                           0.0                                0.0       0.0   
4                           0.0                                0.0       0.0   
...                         ...                                ...       ...   
3424507                     0.0                                0.0       0.0   
34245

In [9]:
historical_weather_df = pd.read_csv('./Data/predict-energy-behavior-of-prosumers/historical_weather.csv')
historical_weather_df

datetime  temperature  dewpoint  rain  snowfall  \
0        2021-09-01 00:00:00         14.2      11.6   0.0       0.0   
1        2021-09-01 00:00:00         13.9      11.5   0.0       0.0   
2        2021-09-01 00:00:00         14.0      12.5   0.0       0.0   
3        2021-09-01 00:00:00         14.6      11.5   0.0       0.0   
4        2021-09-01 00:00:00         15.7      12.9   0.0       0.0   
...                      ...          ...       ...   ...       ...   
1710797  2023-05-30 10:00:00         11.7       4.6   0.0       0.0   
1710798  2023-05-30 10:00:00         12.3       3.5   0.0       0.0   
1710799  2023-05-30 10:00:00          9.8       3.0   0.0       0.0   
1710800  2023-05-30 10:00:00         11.7       1.6   0.0       0.0   
1710801  2023-05-30 10:00:00         12.0       1.4   0.0       0.0   

         surface_pressure  cloudcover_total  cloudcover_low  cloudcover_mid  \
0                  1015.9                31              31               0   
1                  1010.7                33              37               0   
2                  1015.0                31              34               0   
3                  1017.3                 0               0               0   
4                  1014.0                22              25               0   
...                   ...               ...             ...             ...   
1710797            1018.9                40               9              54   
1710798            1019.0                46               4              70   
1710799            1019.2                41               4              62   
1710800            1019.0                44               0              73   
1710801            1016.1                38               0              63   

         cloudcover_high  windspeed_10m  winddirection_10m  \
0                     11       7.083333                  8   
1                      0       5.111111                359   
2                      0       6.333333                355   
3                      0       8.083333                297   
4                      0       8.416667                  5   
...                  ...            ...                ...   
1710797                0       1.055556                253   
1710798                0       0.805556                263   
1710799                0       1.972222                285   
1710800                0       3.500000                307   
1710801                0       3.250000                304   

         shortwave_radiation  direct_solar_radiation  diffuse_radiation  \
0                        0.0                     0.0                0.0   
1                        0.0                     0.0                0.0   
2                        0.0                     0.0                0.0   
3                      358.0                   277.0               81.0   
4                        0.0                     0.0                0.0   
...                      ...                     ...                ...   
1710797                567.0                   392.0              175.0   
1710798                581.0                   407.0              174.0   
1710799                609.0                   432.0              177.0   
1710800                658.0                   521.0              137.0   
1710801                672.0                   550.0              122.0   

         latitude  longitude  data_block_id  
0            57.6       21.7            1.0  
1            57.6       22.2            1.0  
2            57.6       22.7            1.0  
3            57.6       23.2            1.0  
4            57.6       23.7            1.0  
...           ...        ...            ...  
1710797      59.7       26.2          637.0  
1710798      59.7       26.7          637.0  
1710799      59.7       27.2          637.0  
1710800      59.7       27.7          637.0  
1710801      59.7       28.2          637.0  

[1710802 rows x 18 

In [10]:
electricity_prices_df = pd.read_csv("./Data/predict-energy-behavior-of-prosumers/electricity_prices.csv")
electricity_prices_df 

forecast_date  euros_per_mwh          origin_date  data_block_id
0      2021-09-01 00:00:00          92.51  2021-08-31 00:00:00              1
1      2021-09-01 01:00:00          88.90  2021-08-31 01:00:00              1
2      2021-09-01 02:00:00          87.35  2021-08-31 02:00:00              1
3      2021-09-01 03:00:00          86.88  2021-08-31 03:00:00              1
4      2021-09-01 04:00:00          88.43  2021-08-31 04:00:00              1
...                    ...            ...                  ...            ...
15281  2023-05-30 19:00:00          82.10  2023-05-29 19:00:00            637
15282  2023-05-30 20:00:00         150.85  2023-05-29 20:00:00            637
15283  2023-05-30 21:00:00          82.10  2023-05-29 21:00:00            637
15284  2023-05-30 22:00:00          82.09  2023-05-29 22:00:00            637
15285  2023-05-30 23:00:00          -1.29  2023-05-29 23:00:00            637

[15286 rows x 4 columns]

In [11]:
county_id_to_name_map_df = pd.read_json("./Data/predict-energy-behavior-of-prosumers/county_id_to_name_map.json",orient='index')
county_id_to_name_map_df

0
0        HARJUMAA
1         HIIUMAA
2     IDA-VIRUMAA
3        JÄRVAMAA
4       JÕGEVAMAA
5   LÄÄNE-VIRUMAA
6        LÄÄNEMAA
7        PÄRNUMAA
8        PÕLVAMAA
9        RAPLAMAA
10       SAAREMAA
11       TARTUMAA
12        UNKNOWN
13       VALGAMAA
14    VILJANDIMAA
15        VÕRUMAA

In [12]:
# create ids from index
county_id_to_name_map_df['county_ids'] = county_id_to_name_map_df.index
county_id_to_name_map_df

0  county_ids
0        HARJUMAA           0
1         HIIUMAA           1
2     IDA-VIRUMAA           2
3        JÄRVAMAA           3
4       JÕGEVAMAA           4
5   LÄÄNE-VIRUMAA           5
6        LÄÄNEMAA           6
7        PÄRNUMAA           7
8        PÕLVAMAA           8
9        RAPLAMAA           9
10       SAAREMAA          10
11       TARTUMAA          11
12        UNKNOWN          12
13       VALGAMAA          13
14    VILJANDIMAA          14
15        VÕRUMAA          15

In [13]:
# change column name from "0" to "county_names"
county_id_to_name_map_df.rename(columns={0:"county_names"}, inplace=True)

In [14]:
county_id_to_name_map_df.columns

Index(['county_names', 'county_ids'], dtype='object')

In [15]:
weather_station_to_county_mapping_df = pd.read_csv("./Data/predict-energy-behavior-of-prosumers/weather_station_to_county_mapping.csv")
weather_station_to_county_mapping_df

county_name  longitude  latitude  county
0           NaN       21.7      57.6     NaN
1           NaN       21.7      57.9     NaN
2           NaN       21.7      58.2     NaN
3           NaN       21.7      58.5     NaN
4           NaN       21.7      58.8     NaN
..          ...        ...       ...     ...
107         NaN       28.2      58.5     NaN
108         NaN       28.2      58.8     NaN
109         NaN       28.2      59.1     NaN
110         NaN       28.2      59.4     NaN
111         NaN       28.2      59.7     NaN

[112 rows x 4 columns]

<h1>Verify and correct data types and Impute the Null Values</h1>

In [120]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2017824 entries, 0 to 2018351
Data columns (total 9 columns):
 #   Column              Dtype  
---  ------              -----  
 0   county              int64  
 1   is_business         int64  
 2   product_type        int64  
 3   target              float64
 4   is_consumption      int64  
 5   datetime            object 
 6   data_block_id       int64  
 7   row_id              int64  
 8   prediction_unit_id  int64  
dtypes: float64(1), int64(7), object(1)
memory usage: 153.9+ MB


In [16]:
#check null values in train data
train_df.isna().sum()

county                  0
is_business             0
product_type            0
target                528
is_consumption          0
datetime                0
data_block_id           0
row_id                  0
prediction_unit_id      0
dtype: int64

In [17]:
train_df.count()

county                2018352
is_business           2018352
product_type          2018352
target                2017824
is_consumption        2018352
datetime              2018352
data_block_id         2018352
row_id                2018352
prediction_unit_id    2018352
dtype: int64

<h2>Train data has 528 Null values in the target columns in more than 2 million rows we can drop</h2>

In [18]:
train_df.dropna(inplace=True)

In [19]:
#check null values in client data
client_df.isna().sum()

product_type          0
county                0
eic_count             0
installed_capacity    0
is_business           0
date                  0
data_block_id         0
dtype: int64

In [87]:
gas_prices_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 637 entries, 0 to 636
Data columns (total 5 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   forecast_date          637 non-null    object 
 1   lowest_price_per_mwh   637 non-null    float64
 2   highest_price_per_mwh  637 non-null    float64
 3   origin_date            637 non-null    object 
 4   data_block_id          637 non-null    int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 25.0+ KB


In [20]:
#check null values in gas_prices data
gas_prices_df.isna().sum()

forecast_date            0
lowest_price_per_mwh     0
highest_price_per_mwh    0
origin_date              0
data_block_id            0
dtype: int64

In [127]:
gas_prices_df[["forecast_date", "origin_date"]] = gas_prices_df[["forecast_date", "origin_date"]].apply(pd.to_datetime)

In [130]:
gas_prices_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 637 entries, 0 to 636
Data columns (total 5 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   forecast_date          637 non-null    datetime64[ns]
 1   lowest_price_per_mwh   637 non-null    float64       
 2   highest_price_per_mwh  637 non-null    float64       
 3   origin_date            637 non-null    datetime64[ns]
 4   data_block_id          637 non-null    int64         
dtypes: datetime64[ns](2), float64(2), int64(1)
memory usage: 25.0 KB


In [124]:
electricity_prices_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15286 entries, 0 to 15285
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   forecast_date  15286 non-null  object 
 1   euros_per_mwh  15286 non-null  float64
 2   origin_date    15286 non-null  object 
 3   data_block_id  15286 non-null  int64  
dtypes: float64(1), int64(1), object(2)
memory usage: 477.8+ KB


In [21]:
#check null values in electricity_prices data
electricity_prices_df.isna().sum()

forecast_date    0
euros_per_mwh    0
origin_date      0
data_block_id    0
dtype: int64

In [89]:
forecast_weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3424512 entries, 0 to 3424511
Data columns (total 18 columns):
 #   Column                             Dtype  
---  ------                             -----  
 0   latitude                           float64
 1   longitude                          float64
 2   origin_datetime                    object 
 3   hours_ahead                        int64  
 4   temperature                        float64
 5   dewpoint                           float64
 6   cloudcover_high                    float64
 7   cloudcover_low                     float64
 8   cloudcover_mid                     float64
 9   cloudcover_total                   float64
 10  10_metre_u_wind_component          float64
 11  10_metre_v_wind_component          float64
 12  data_block_id                      int64  
 13  forecast_datetime                  object 
 14  direct_solar_radiation             float64
 15  surface_solar_radiation_downwards  float64
 16  snowfall          

In [22]:
#check null values in forecast_weather data
forecast_weather_df.isna().sum()

latitude                             0
longitude                            0
origin_datetime                      0
hours_ahead                          0
temperature                          0
dewpoint                             0
cloudcover_high                      0
cloudcover_low                       0
cloudcover_mid                       0
cloudcover_total                     0
10_metre_u_wind_component            0
10_metre_v_wind_component            0
data_block_id                        0
forecast_datetime                    0
direct_solar_radiation               0
surface_solar_radiation_downwards    2
snowfall                             0
total_precipitation                  0
dtype: int64

In [90]:
historical_weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1710802 entries, 0 to 1710801
Data columns (total 18 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   datetime                object 
 1   temperature             float64
 2   dewpoint                float64
 3   rain                    float64
 4   snowfall                float64
 5   surface_pressure        float64
 6   cloudcover_total        int64  
 7   cloudcover_low          int64  
 8   cloudcover_mid          int64  
 9   cloudcover_high         int64  
 10  windspeed_10m           float64
 11  winddirection_10m       int64  
 12  shortwave_radiation     float64
 13  direct_solar_radiation  float64
 14  diffuse_radiation       float64
 15  latitude                float64
 16  longitude               float64
 17  data_block_id           float64
dtypes: float64(12), int64(5), object(1)
memory usage: 234.9+ MB


In [23]:
#check null values in historical_weather data
historical_weather_df.isna().sum()

datetime                  0
temperature               0
dewpoint                  0
rain                      0
snowfall                  0
surface_pressure          0
cloudcover_total          0
cloudcover_low            0
cloudcover_mid            0
cloudcover_high           0
windspeed_10m             0
winddirection_10m         0
shortwave_radiation       0
direct_solar_radiation    0
diffuse_radiation         0
latitude                  0
longitude                 0
data_block_id             0
dtype: int64

In [91]:
weather_station_to_county_mapping_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112 entries, 0 to 111
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   county_name  49 non-null     object 
 1   longitude    112 non-null    float64
 2   latitude     112 non-null    float64
 3   county       49 non-null     float64
dtypes: float64(3), object(1)
memory usage: 3.6+ KB


In [24]:
#check null values in weather_station_to_county_mapping data
weather_station_to_county_mapping_df.isna().sum()

county_name    63
longitude       0
latitude        0
county         63
dtype: int64

In [25]:
weather_station_to_county_mapping_df.count()

county_name     49
longitude      112
latitude       112
county          49
dtype: int64

<h2> Weather station to county mapping data has 63 Null values in the county name and county columns in 112 values. Match the unknown county values to their closest counties based on longitute and latitude </h2>

In [34]:
location_county_no_nulls = weather_station_to_county_mapping_df.dropna(how='any')
location_county_no_nulls

county_name  longitude  latitude  county
10        Saaremaa       22.2      58.2    10.0
11        Saaremaa       22.2      58.5    10.0
19        Saaremaa       22.7      58.5    10.0
20         Hiiumaa       22.7      58.8     1.0
27        Saaremaa       23.2      58.5    10.0
35        Pärnumaa       23.7      58.5     7.0
36        Läänemaa       23.7      58.8     6.0
37        Läänemaa       23.7      59.1     6.0
43        Pärnumaa       24.2      58.5     7.0
44        Raplamaa       24.2      58.8     9.0
45        Harjumaa       24.2      59.1     0.0
46        Harjumaa       24.2      59.4     0.0
50        Pärnumaa       24.7      58.2     7.0
51        Pärnumaa       24.7      58.5     7.0
52        Raplamaa       24.7      58.8     9.0
53        Raplamaa       24.7      59.1     9.0
54        Harjumaa       24.7      59.4     0.0
58        Pärnumaa       25.2      58.2     7.0
59     Viljandimaa       25.2      58.5    14.0
60        Järvamaa       25.2      58.8     3.0
61        Harjumaa       25.2      59.1     0.0
62        Harjumaa       25.2      59.4     0.0
66     Viljandimaa       25.7      58.2    14.0
67     Viljandimaa       25.7      58.5    14.0
68        Järvamaa       25.7      58.8     3.0
69        Järvamaa       25.7      59.1     3.0
70        Harjumaa       25.7      59.4     0.0
73        Valgamaa       26.2      57.9    13.0
74        Tartumaa       26.2      58.2    11.0
75       Jõgevamaa       26.2      58.5     4.0
76       Jõgevamaa       26.2      58.8     4.0
77   Lääne-Virumaa       26.2      59.1     5.0
78   Lääne-Virumaa       26.2      59.4     5.0
80         Võrumaa       26.7      57.6    15.0
81         Võrumaa       26.7      57.9    15.0
82        Tartumaa       26.7      58.2    11.0
83        Tartumaa       26.7      58.5    11.0
84       Jõgevamaa       26.7      58.8     4.0
85   Lääne-Virumaa       26.7      59.1     5.0
86   Lääne-Virumaa       26.7      59.4     5.0
88         Võrumaa       27.2      57.6    15.0
89         Võrumaa       27.2      57.9    15.0
90        Põlvamaa       27.2      58.2     8.0
91        Tartumaa       27.2      58.5    11.0
93     Ida-Virumaa       27.2      59.1     2.0
94     Ida-Virumaa       27.2      59.4     2.0
97         Võrumaa       27.7      57.9    15.0
101    Ida-Virumaa       27.7      59.1     2.0
102    Ida-Virumaa       27.7      59.4     2.0

In [35]:
labeled_locations = list(location_county_no_nulls.itertuples(index=False))
labeled_locations

[Pandas(county_name='Saaremaa', longitude=22.2, latitude=58.2, county=10.0),
 Pandas(county_name='Saaremaa', longitude=22.2, latitude=58.49999999999999, county=10.0),
 Pandas(county_name='Saaremaa', longitude=22.7, latitude=58.49999999999999, county=10.0),
 Pandas(county_name='Hiiumaa', longitude=22.7, latitude=58.79999999999999, county=1.0),
 Pandas(county_name='Saaremaa', longitude=23.2, latitude=58.49999999999999, county=10.0),
 Pandas(county_name='Pärnumaa', longitude=23.7, latitude=58.49999999999999, county=7.0),
 Pandas(county_name='Läänemaa', longitude=23.7, latitude=58.79999999999999, county=6.0),
 Pandas(county_name='Läänemaa', longitude=23.7, latitude=59.09999999999999, county=6.0),
 Pandas(county_name='Pärnumaa', longitude=24.2, latitude=58.49999999999999, county=7.0),
 Pandas(county_name='Raplamaa', longitude=24.2, latitude=58.79999999999999, county=9.0),
 Pandas(county_name='Harjumaa', longitude=24.2, latitude=59.09999999999999, county=0.0),
 Pandas(county_name='Harjumaa',

In [39]:
  weather_station_to_county_mapping_df[weather_station_to_county_mapping_df.county.isnull()]

county_name  longitude  latitude  county
0           NaN       21.7      57.6     NaN
1           NaN       21.7      57.9     NaN
2           NaN       21.7      58.2     NaN
3           NaN       21.7      58.5     NaN
4           NaN       21.7      58.8     NaN
..          ...        ...       ...     ...
107         NaN       28.2      58.5     NaN
108         NaN       28.2      58.8     NaN
109         NaN       28.2      59.1     NaN
110         NaN       28.2      59.4     NaN
111         NaN       28.2      59.7     NaN

[63 rows x 4 columns]

In [44]:
location_county_nulls = weather_station_to_county_mapping_df[weather_station_to_county_mapping_df.county.isnull()][['longitude', 'latitude']]
location_county_nulls

longitude  latitude
0         21.7      57.6
1         21.7      57.9
2         21.7      58.2
3         21.7      58.5
4         21.7      58.8
..         ...       ...
107       28.2      58.5
108       28.2      58.8
109       28.2      59.1
110       28.2      59.4
111       28.2      59.7

[63 rows x 2 columns]

In [45]:
unlabeled_locations = list(location_county_nulls.itertuples(index=False))
unlabeled_locations

[Pandas(longitude=21.7, latitude=57.6),
 Pandas(longitude=21.7, latitude=57.9),
 Pandas(longitude=21.7, latitude=58.2),
 Pandas(longitude=21.7, latitude=58.49999999999999),
 Pandas(longitude=21.7, latitude=58.79999999999999),
 Pandas(longitude=21.7, latitude=59.09999999999999),
 Pandas(longitude=21.7, latitude=59.399999999999984),
 Pandas(longitude=21.7, latitude=59.69999999999998),
 Pandas(longitude=22.2, latitude=57.6),
 Pandas(longitude=22.2, latitude=57.9),
 Pandas(longitude=22.2, latitude=58.79999999999999),
 Pandas(longitude=22.2, latitude=59.09999999999999),
 Pandas(longitude=22.2, latitude=59.399999999999984),
 Pandas(longitude=22.2, latitude=59.69999999999998),
 Pandas(longitude=22.7, latitude=57.6),
 Pandas(longitude=22.7, latitude=57.9),
 Pandas(longitude=22.7, latitude=58.2),
 Pandas(longitude=22.7, latitude=59.09999999999999),
 Pandas(longitude=22.7, latitude=59.399999999999984),
 Pandas(longitude=22.7, latitude=59.69999999999998),
 Pandas(longitude=23.2, latitude=57.6),
 

In [69]:
def find_closest_label(location, labeled_locations):
    closest_label = None
    min_distance = float('inf')

    for labeled_location in labeled_locations:
        distance = geodesic(location, (labeled_location.latitude, labeled_location.longitude)).kilometers
        if distance < min_distance:
            min_distance = distance
            closest_county = labeled_location.county
            closest_county_name = labeled_location.county_name

    return closest_county, closest_county_name

In [74]:
assigned_counties = []

for location in unlabeled_locations:
    county, county_name = find_closest_label((location.latitude, location.longitude), labeled_locations)
    assigned_counties.append((county_name, location.latitude, location.longitude, county))
    print(f"Location {location} is closest to {county_name}, county number {county}")

Location Pandas(longitude=21.7, latitude=57.6) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=21.7, latitude=57.9) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=21.7, latitude=58.2) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=21.7, latitude=58.49999999999999) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=21.7, latitude=58.79999999999999) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=21.7, latitude=59.09999999999999) is closest to Hiiumaa, county number 1.0
Location Pandas(longitude=21.7, latitude=59.399999999999984) is closest to Hiiumaa, county number 1.0
Location Pandas(longitude=21.7, latitude=59.69999999999998) is closest to Hiiumaa, county number 1.0
Location Pandas(longitude=22.2, latitude=57.6) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=22.2, latitude=57.9) is closest to Saaremaa, county number 10.0
Location Pandas(longitude=22.2, latitude

In [79]:
assigned_counties_df = pd.DataFrame(assigned_counties, columns = ['county_name', 'longitude', 'latitude', 'county'])
assigned_counties_df

county_name  longitude  latitude  county
0      Saaremaa       57.6      21.7    10.0
1      Saaremaa       57.9      21.7    10.0
2      Saaremaa       58.2      21.7    10.0
3      Saaremaa       58.5      21.7    10.0
4      Saaremaa       58.8      21.7    10.0
..          ...        ...       ...     ...
58     Tartumaa       58.5      28.2    11.0
59  Ida-Virumaa       58.8      28.2     2.0
60  Ida-Virumaa       59.1      28.2     2.0
61  Ida-Virumaa       59.4      28.2     2.0
62  Ida-Virumaa       59.7      28.2     2.0

[63 rows x 4 columns]

In [83]:
# create a new dataframe using imputed county name / county if values
weather_station_to_county_mapping_df_imputed = pd.concat([location_county_no_nulls, assigned_counties_df])
weather_station_to_county_mapping_df_imputed

county_name  longitude  latitude  county
10     Saaremaa       22.2      58.2    10.0
11     Saaremaa       22.2      58.5    10.0
19     Saaremaa       22.7      58.5    10.0
20      Hiiumaa       22.7      58.8     1.0
27     Saaremaa       23.2      58.5    10.0
..          ...        ...       ...     ...
58     Tartumaa       58.5      28.2    11.0
59  Ida-Virumaa       58.8      28.2     2.0
60  Ida-Virumaa       59.1      28.2     2.0
61  Ida-Virumaa       59.4      28.2     2.0
62  Ida-Virumaa       59.7      28.2     2.0

[112 rows x 4 columns]

<h2> Cast county data type to int as it refers to county id </h2>

In [98]:
weather_station_to_county_mapping_df_imputed = weather_station_to_county_mapping_df_imputed.astype({'county': 'int'})

In [99]:
weather_station_to_county_mapping_df_imputed.dtypes

county_name     object
longitude      float64
latitude       float64
county           int64
dtype: object

In [100]:
weather_station_to_county_mapping_df_imputed

county_name  longitude  latitude  county
10     Saaremaa       22.2      58.2      10
11     Saaremaa       22.2      58.5      10
19     Saaremaa       22.7      58.5      10
20      Hiiumaa       22.7      58.8       1
27     Saaremaa       23.2      58.5      10
..          ...        ...       ...     ...
58     Tartumaa       58.5      28.2      11
59  Ida-Virumaa       58.8      28.2       2
60  Ida-Virumaa       59.1      28.2       2
61  Ida-Virumaa       59.4      28.2       2
62  Ida-Virumaa       59.7      28.2       2

[112 rows x 4 columns]

<h1> Now we imputed the null values, explore and visualize the data </h1>

In [104]:
gas_prices_df

forecast_date  lowest_price_per_mwh  highest_price_per_mwh origin_date  \
0      2021-09-01                 45.23                  46.32  2021-08-31   
1      2021-09-02                 45.62                  46.29  2021-09-01   
2      2021-09-03                 45.85                  46.40  2021-09-02   
3      2021-09-04                 46.30                  46.80  2021-09-03   
4      2021-09-05                 46.30                  46.58  2021-09-04   
..            ...                   ...                    ...         ...   
632    2023-05-26                 29.10                  34.10  2023-05-25   
633    2023-05-27                 28.30                  34.10  2023-05-26   
634    2023-05-28                 28.10                  34.10  2023-05-27   
635    2023-05-29                 28.16                  36.98  2023-05-28   
636    2023-05-30                 29.00                  34.00  2023-05-29   

     data_block_id  
0                1  
1                2  
2                3  
3                4  
4                5  
..             ...  
632            633  
633            634  
634            635  
635            636  
636            637  

[637 rows x 5 columns]

In [131]:
gas_prices_df.groupby('data_block_id').count()

forecast_date  lowest_price_per_mwh  highest_price_per_mwh  \
data_block_id                                                               
1                          1                     1                      1   
2                          1                     1                      1   
3                          1                     1                      1   
4                          1                     1                      1   
5                          1                     1                      1   
...                      ...                   ...                    ...   
633                        1                     1                      1   
634                        1                     1                      1   
635                        1                     1                      1   
636                        1                     1                      1   
637                        1                     1                      1   

               origin_date  
data_block_id               
1                        1  
2                        1  
3                        1  
4                        1  
5                        1  
...                    ...  
633                      1  
634                      1  
635                      1  
636                      1  
637                      1  

[637 rows x 4 columns]

In [132]:
gas_prices_df.dtypes

forecast_date            datetime64[ns]
lowest_price_per_mwh            float64
highest_price_per_mwh           float64
origin_date              datetime64[ns]
data_block_id                     int64
dtype: object

In [138]:
# see if forecast date is always 1 day ahead of the origin date
(gas_prices_df['origin_date'] + pd.Timedelta(days=1) == gas_prices_df['forecast_date']).eq(True).all()

True

In [141]:
gas_prices_df['origin_date'].nunique()

637

<h2> So, gas prices dataset has 637 unique block id, origin date and forecast date </h2>